web/ui/src/features/component/ComponentGraph.tsx

```diff
@@ -0,0 +1,394 @@
+  h: number;
+}
+
+export interface ComponentGraphProps {
+  components: ComponentInfo[];
+}
+
+/**
+ * ComponentGraph renders an SVG with relationships between defined components.
+ * The components prop must be a non-empty array.
+ */
+export const ComponentGraph: FC<ComponentGraphProps> = (props) => {
+  const baseComponentPath = useHref('/component');
+  const svgRef = useRef<SVGSVGElement>(null);
+
+  useEffect(() => {
+    // NOTE(rfratto): The default units of svg are in pixels.
+
+    const [nodeWidth, nodeHeight] = [150, 75];
+    const nodeMargin = 25;
+    const nodePadding = 5;
+
+    const contentHeight = nodeHeight - nodePadding * 2;
+
+    const widthCache: Record<string, number> = {
+      foo: 5,
+    };
+
+    const builder = dagStratify()
+      .id<IdOperator<ComponentInfo>>((n) => n.id)
+      .parentIds<ParentIdsOperator<ComponentInfo>>((n) => n.referencedBy);
+    const dag = builder(props.components);
+
+    const layout = sugiyama()
+      .decross(decrossOpt())
+      .coord(coordQuad())
+      .nodeSize<NodeSizeAccessor<HasId & HasParentIds, undefined>>((n) => {
+        // nodeSize is the full amount of space you want the node to take up.
+        //
+        // It can be considered similar to the box model: margin and padding should
+        // be added to the size here;
+
+        if (n === undefined) {
+          return [nodeWidth + nodeMargin + nodePadding * 2, nodeHeight + nodeMargin + nodePadding * 2];
+        }
+
+        // Calculate how much width the text needs to be displayed.
+        let width = nodeWidth;
+
+        const textWidth = calcTextWidth(n.data.id, "bold 13px 'Roboto', sans-serif");
+        if (textWidth != null && textWidth > width) {
+          width = textWidth;
+        }
+
+        // Cache the width so it can be used while plotting the SVG.
+        widthCache[n.data.id] = width;
+
+        return [width + nodeMargin + nodePadding * 2, nodeHeight + nodeMargin + nodePadding * 2];
+      });
+    const { width, height } = layout(dag);
+
+    // svgRef.current needs to be cast to an Element for type checks to work.
+    // SVGSVGElement doesn't extend element and prevents zoom from
+    // typechecking.
+    //
+    // Everything still seems to work even with the type cast.
+    const svgSelection = d3.select(svgRef.current as Element);
+    svgSelection.selectAll('*').remove(); // Clear svg content before adding new elements
+    svgSelection.attr('viewBox', [0, 0, width, height].join(' '));
+
+    const svgWrapper = svgSelection.append('g');
+
+    const zoom = d3Zoom
+      .zoom()
+      .scaleExtent([0.1, 3])
+      .on('zoom', (e) => {
+        svgWrapper.attr('transform', e.transform);
+      });
+
+    svgSelection.call(zoom).call(zoom.transform, d3Zoom.zoomIdentity);
+
+    // Add a marker element so we can draw an arrow pointing between nodes.
+    svgWrapper
+      .append('defs')
+      .append('marker')
+      .attr('id', 'arrow')
+      .attr('viewBox', [0, 0, 20, 20])
```

Comment:
What are the 0,0,20,20 referenced a few times? Are those the dimensions for the box with the component name?

Reply:
Yeah, this assigns a [viewBox attribute](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/viewBox) with x=0, y=0, w=20, h=20. It's basically the viewport size where elements can be drawn.

Do you see 0,0,20,20 anywhere else? I only see the one instance here

Reply:
Ten or so lines down that I assume is drawing the border

Reply:
Oh, that's slightly different. That's the actual arrow shape. The 0,0,20,20 is the actual box the arrow shape can be drawn in. 


I added some comments to try to clear it up, is it better now?